### PR TITLE
Deprecate MarkerView

### PR DESCRIPTION
**Pull request by tobrun**
This PR deprecates the usage of MarkerView and replaces those use-cases with using a `SymbolLayer` + `SymbolGenerator`. This allows to benefit from the options that symbolLayer exposes while using any Android SDK View as `iconImage`.

 - First commit updates naming of an older example (no refs to sprite in public api of the android binding)
 - Second commit show cases SymbolGenerator (see gif below)
 - Third commit deprecates MarkerView API

ezgif com-video-to-gif 11

cc @mapbox/android

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java [hunk 1]
```diff
@@ -14,7 +18,15 @@
  *
  * @param <U> Type of the marker view to be composed.
  * @param <T> Type of the builder to be used for composing.
+ * @deprecated Use {@link com.mapbox.mapboxsdk.style.utils.SymbolGenerator} in combination with
+ * {@link com.mapbox.mapboxsdk.style.layers.SymbolLayer} instead. Add the result from
+ * {@link com.mapbox.mapboxsdk.style.utils.SymbolGenerator#generate(View)} with
+ * {@link MapboxMap#addImage(String, Bitmap)} and link the id of the image with
+ * {@link com.mapbox.mapboxsdk.style.layers.PropertyFactory#iconImage(String)} when setting properties to a SymbolLayer
+ * with {@link com.mapbox.mapboxsdk.style.layers.SymbolLayer#setProperties(PropertyValue[])} or
+ * {@link com.mapbox.mapboxsdk.style.layers.SymbolLayer#withProperties(PropertyValue[])}.
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java]: Could we add a link to an example using the new proposed approach? Explaining the general approach is great but I think that some working code would make the transition easier.

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -50,7 +50,8 @@
     <string name="activity_query_rendered_features_box_highlight">Highlight features in box</string>
     <string name="activity_query_source_features">Query source features</string>
     <string name="activity_symbol_layer">Symbols</string>
-    <string name="activity_add_sprite">Add Custom Sprite</string>
+    <string name="activity_add_symbol">Add Custom Symbol</string>
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDKTestApp/src/main/res/values/strings.xml]: Maybe add a note saying "replaces MarkerView approach"?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java (hunk 1 above)
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java]: Maybe `CustomSymbolLayerActivity` or `SymbolGeneratorActivity` if they're reference implementations for the new approach?

**Comment by 1ec5** [comment]
Can you speak to the performance of SymbolGenerator compared to MarkerViews with 100, 500, 1,000 markers on the map?

**Comment by tobrun** [comment]
MarkerViews are Android SDK views synchronized above our rendering surface (currently SurfaceView, optional to switch to TextureView for increased view synchronization performance). What happens for a MarkerView is that the user inflates a View and we will synchronize its position to a LatLng every time the map changes. This synchronization will result in measure and layout phases with every change. 

With SymbolGenerator we measure the view once and create a bitmap of it, no performance hit while using the map. One difference though is that MarkerView are inflated on the fly but with SymbolGenerator you will be creating your views/bitmaps up front. I'm not really seeing this as an issue as this is the same ux as with SymbolLayer. 

I will look into

**Comment by tobrun** [comment]
I updated the example to create a View and SymbolLayer for each Feature in the GeoJSON (37 features):

ezgif com-video-to-gif 13

I would say this example would show a lot of jiggling when implemented with MarkerViews.

**Comment by tobrun** [comment]
Did another test with this [geojson](https://d2ad6b4ur7yvpq.cloudfront.net/naturalearth-3.3.0/ne_10m_ports.geojson). This resembles ports in the world and consists out of 1081 points. 

ezgif com-video-to-gif 14

You can see this is not acceptable in terms of performance (tested on low end device) though the same use-case implemented with `textField` shows similar issues:

ezgif com-video-to-gif 15

**Comment by tobrun** [comment]
for reference here is a gif with 100 and 1000 markerviews:

ezgif com-video-to-gif 17

Adding 1000 resutls in an ANR.

**Comment by 1ec5** [comment]
Thanks for the analysis. So would this be a good characterization of the tradeoff?

MarkerViews:

* Support views that change dynamically with animation – doing that performantly would require #6177
* More performant for markers that move frequently with animation (think cars moving around atop the map)

SymbolGenerator:

* Full runtime styling support
* More performant for markers with static coordinates and static content (no jiggling and smaller memory footprint)
* Markers correctly scale (#5169) and skew (#5218) when the map is tilted

What about some of the other behaviors listed in #6181?

* Can a SymbolGenerator persist beyond the lifetime of a style (#6180)?
* If the developer wants to implement interactivity, can they use feature querying and rely on the feature IDs to correspond to view IDs (#6178 #7376)?

**Comment by tobrun** [comment]
Capturing from internal conversations that we are planning to put the SymbolGenerator code inside a utils plugin instead of the SDK. I'm going to rework this PR to only deprecate MarkerView and point the deprecation notices to use SymbolLayer instead.

**Comment by tobrun** [comment]
with https://github.com/mapbox/mapbox-gl-native/pull/9782/commits/ed2ae27b5c0632c89319df13c1859f1186217378 the suggestion from https://github.com/mapbox/mapbox-gl-native/pull/9782#issuecomment-329093796 has been applied. 

cc @zugaldia

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java [hunk 3]
```diff
@@ -14,7 +14,10 @@
  *
  * @param <U> Type of the marker view to be composed.
  * @param <T> Type of the builder to be used for composing.
+ * @deprecated Use a {@link com.mapbox.mapboxsdk.style.layers.SymbolLayer} instead. An example of converting Android
+ * SDK views to be used as a symbol see https://gist.github.com/tobrun/349777a6ef0dfb55245a544344ee197a.
```
Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java]: How about we instead include this approach as an activity in the TestApp while we work on making it available as a plugin?

Comment (tobrun) [reply platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java]: I added SymbolGeneratorActivity to the testapp, replaced all references to the gist with a permalink to an Activity in the testapp.

Comment (zugaldia) [reply platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java]: @tobrun 👍

Comment (tobrun) [reply platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerViewOptions.java]: If you are on-board with this change, could you re-review?

**Comment by joaopedro1520** [comment]
Hello I have been using marker views and have a lot of issues with performance since i need to add a lot of markers to do map. It takes a while to draw them and the device becomes unresponsive, but after it's done the map works well with gestures and moving around.

I already opened a issue [#9726](https://github.com/mapbox/mapbox-gl-native/issues/9726) and got the suggestion to try symbol layer as a way to improve performance.
But from my testing adding ~500 markers the map starts being really slow and gestures have huge lag.
Here is my implementation to add the markers, am I doing anything wrong? 
Thanks 

`                    for (MapObject tmpObject : data) {

                        Random rand = new Random();

                        FeatureCollection markers = FeatureCollection.fromFeatures(new Feature[] {Feature.fromGeometry(Point.fromCoordinates(new double[]{tmpObject.getLongitude(), tmpObject.getLatitude()}))});
                        int n = rand.nextInt(5000000) + 1;

                        Source source = new GeoJsonSource(GENERATED_SYMBOL_SOURCE_ID + n, markers);
                        mMapboxMap.addSource(source);

                        ImageView imageView = new ImageView(getAppCtx());
                        imageView.setImageResource(R.drawable.mapbox_marker_icon_default);


                        // create bitmap from view
                        String iconId = GENERATED_SYMBOL_IMAGE_ID_BASE + tmpObject.getId();
                        mMapboxMap.addImage(iconId,SymbolGenerator.generate(imageView) );

                        // create layer for bitmap and filter source on name
                        String layerId = GENERATED_SYMBOL_LAYER_ID_BASE + tmpObject.getId();
                        layerIds.add(layerId);
                        mMapboxMap.addLayer(new SymbolLayer(layerId, GENERATED_SYMBOL_SOURCE_ID+n)
                                .withProperties(
                                        iconImage(iconId),
                                        iconAllowOverlap(false)
                                )
                        );


                    }`

**Comment by tobrun** [comment]
updated the example to use 1 layer for all different used symbols in 
https://github.com/mapbox/mapbox-gl-native/pull/9782/commits/68f32bc104422207c64da8d90e8411b138d87f04

**Comment by josileudo** [comment]
COMO EU FARAI PARA ADICIONAR MARCAÇÕES E SALVAR ESSAS MARCAÇÕES DENTRO DE UM BANCO DE DADOS OU GeoJson?

**Pull request merged**